ubuntu_image/builder.py

```diff
@@ -403,53 +402,38 @@ def make_disk(self):
             image = Image(self.disk_img, self.image_size)
         offset_writes = []
         part_offsets = {}
-        root_offset = 0
         for i, part in enumerate(volume.structures):
             if part.name is not None:
                 part_offsets[part.name] = part.offset
             if part.offset_write is not None:
                 offset_writes.append((part.offset, part.offset_write))
-            image.copy_blob(self.boot_images[i],
+            image.copy_blob(self.part_images[i],
                             bs='1M', seek=part.offset // MiB(1),
                             count=ceil(part.size / MiB(1)),
                             conv='notrunc')
-            if part.type in ('bare', 'mbr'):
+            if (part.role is StructureRole.mbr or
+                    part.type == 'bare'):
                 continue
             # sgdisk takes either a sector or a KiB/MiB argument; assume
             # that the offset and size are always multiples of 1MiB.
             #
             # XXX Size must not be zero, which will happen if part.size < 1MiB
             partition_args = dict(
-                new='{}M:+{}M'.format(
-                    part.offset // MiB(1), part.size // MiB(1)),
+                new='{}M:+{}K'.format(
+                    part.offset // MiB(1), ceil(part.size / 1024)),
```

Comment:
TL;DR: +1

I just saw [LP: #1460715](https://bugs.launchpad.net/ubuntu-image/+bug/1460715) today, so I think that's an opportune change, although we might also need to change the first 'M' (I haven't fully reviewed the impact of that bug on u-i yet).  Thanks for the explanation.

Noting that there's a subtle behavioral difference here.  If part.size is some fraction of a MiB bigger than 1MiB, the partition will only be the floor of the size in MiB under the old code.  So I think the ceil() calculation is more right because the resulting partition may be slightly bigger than the part size, which is okay because it definitely won't be smaller!